### PR TITLE
Summon - next iteration

### PR DESCRIPTION
**Pull request by dmpetrov**
An iteration on summoning (see https://github.com/iterative/dvcx/pull/4):
1. Some minor Summon file format changes and default file name change to `Summon.yaml`.
2. Separating reading objects and pulling data. It is needed for flexibility and especially for data objects with several files like a compound CSV file (see in #2719).
3. Summon object update `update_dobj()`. This is not completed but it does not break current read-only logic.
4. Moving part of the logic to `SummonDesc` class.

* [X] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [X] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [X] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/api.py [hunk 1]
```diff
@@ -13,12 +13,15 @@
 from dvc.exceptions import DvcException, NotDvcRepoError
 from dvc.external_repo import external_repo
 
+DEF_SUMMON = "Summon.yaml"
```
Comment (shcheklein) [inline dvc/api.py]: minor: make it lower case? :) the only example is Makefile? and obviously, not the best example ...

Comment (dmpetrov) [reply dvc/api.py]: Agree. But then we need to consider renaming default repro goal - `Dvcfile`. Or accept two default options `Dvcfile` and `dvcfile`.

Comment (Suor) [reply dvc/api.py]: @dmpetrov it's `Dockerfile`, but `docker-compose.yml`. Basically everything is lowercase if it doesn't resemble Makefiles. Also, why not `dvcsummon.yaml`?

Comment (efiop) [reply dvc/api.py]: I agree with @Suor , plus we have `.dvcignore` already as an example of dvc-specific file :slightly_smiling_face:

File: dvc/api.py [hunk 2]
```diff
@@ -13,12 +13,15 @@
...
+DOBJ_SECTION = "d-objects"
```
Comment (Suor) [inline dvc/api.py]: What does it stand for? Also, we was trying to go away from string constants) This makes it more readable.

File: dvc/api.py [hunk 3]
```diff
@@ -120,94 +127,134 @@ def _make_repo(repo_url=None, rev=None):
...
 
         return objects[0]
 
-    except FileNotFoundError as exc:
-        raise SummonError("Summon file not found") from exc
-    except ruamel.yaml.YAMLError as exc:
-        raise SummonError("Failed to parse summon file") from exc
-    except Invalid as exc:
-        raise SummonError(str(exc)) from exc
+    def update_dobj(self, new_dobj, overwrite=False):
```
Comment (Suor) [inline dvc/api.py]: "update" verb presumes override, imho. Should this be `.set_dobject()`?

File: dvc/api.py [hunk 4]
```diff
@@ -120,94 +127,134 @@ def _make_repo(repo_url=None, rev=None):
...
+        try:
+            name = new_dobj["name"]
+            dobj = self.get_dobject(name)
+
+            if overwrite:
+                idx = self.content[DOBJ_SECTION].index(dobj)
+                self.content[DOBJ_SECTION][idx] = new_dobj
```
Comment (Suor) [inline dvc/api.py]: Maybe we should use dict with name as key instead of a list.

File: dvc/api.py [hunk 5]
```diff
@@ -120,94 +127,134 @@ def _make_repo(repo_url=None, rev=None):
...
+        desc.pull(dobj)
         _args = {**summon_dict.get("args", {}), **(args or {})}
         return _invoke_method(summon_dict["call"], _args, desc.repo.root_dir)
 
 
-@contextmanager
-def prepare_summon(name, repo=None, rev=None, summon_file="dvcsummon.yaml"):
-    """Does a couple of things every summon needs as a prerequisite:
-    clones the repo, parses the summon file and pulls the deps.
+class SummonDesc(object):
```
Comment (Suor) [inline dvc/api.py]: So it's now `SummonFile`. Why do you want to separate it anyway? Will we use single file to work with several objects?

File: dvc/api.py (hunk 2 above)
Comment (dmpetrov) [reply dvc/api.py]: `dvc-objects` or `data-objects`. I'm making it `dvc-objects`.

File: dvc/api.py (hunk 3 above)
Comment (dmpetrov) [reply dvc/api.py]: Right. Changing the `overwrite=False` to `True`.

File: dvc/api.py (hunk 5 above)
Comment (dmpetrov) [reply dvc/api.py]: Changed to `SummonFile`.

Sure, in the resnet example we had many resnet models in a single file. `pretrained-models.pytorch` project contains 50+ models. We probably don't what to have 50 summon files (the code is localized in few files only for all the models).

File: dvc/api.py (hunk 1 above)
Comment (dmpetrov) [reply dvc/api.py]: Let's keep `dvcsummon.yaml`

File: dvc/api.py (hunk 4 above)
Comment (dmpetrov) [reply dvc/api.py]: Gush... great idea.

**Comment by dmpetrov** [comment]
@Suor should we close this one?

**Pull request merged**